### PR TITLE
Integrate GraphOS

### PR DESCRIPTION
**Pull request by michael-watson**
This PR does a number of things:

- Moves backend to separate infrastructure pieces hosted on Railway/Netflify
  - Subgraphs folder where two subgraphs can be independently stared
- Website is now hosted on Netlify
- Modified root `package.json` to have a `start` and `setup` script for local development that includes `rover dev`
- GraphOS schema checks and linting happen on PRs
- GraphOS schema pipeline updates on merge

### CODE REVIEW

File: subgraphs/playback/src/resolvers/Subscription.ts [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
+import { SubscriptionResolvers } from "../__generated__/resolvers-types";
+import { Spotify } from "../dataSources/spotify.types";
+import { TOPICS } from "../utils/constants";
+import { map, distinctUntilChanged } from "rxjs";
+import { equal } from "@wry/equality";
+import { GraphQLResolveInfo } from "graphql";
+import { omit } from "lodash";
+import { PartialDeep } from "type-fest";
+import { selectsField } from "../utils/graphql";
+import { createPlaybackStateObservable } from "../observables";
```
Comment (jerelmiller) [inline subgraphs/playback/src/resolvers/Subscription.ts:10]: Looks like the prettier config changed or something. Mind if we use single quotes for prettier config?

File: subgraphs/playback/src/resolvers/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// exp0ort function itself<TParent extends ParentTypes>(): Resolver<
+//   TParent,
+//   TParent
+// > {
+//   return (parent: TParent) => parent;
+// }
+
+// export function wrapWithSynthetics<TResult, TParent = {}, TArgs = {}>(
```
Comment (jerelmiller) [inline subgraphs/playback/src/resolvers/helpers.ts:23]: This one is super important as it gives us the ability to artificially slow down the query (also necessary for the talk Alessia and I are giving later this year!!). Could we make sure all the resolvers are wrapped with this?

File: subgraphs/playback/src/resolvers/index.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  // Episode,
+  Mutation,
+  PlaybackContext,
+  PlaybackContextType,
+  PlaybackItem,
+  PlaybackState,
+  // PlaylistTrack,
+  RepeatMode,
+  Subscription,
+  // Track
```
Comment (jerelmiller) [inline subgraphs/playback/src/resolvers/index.ts:28]: I'm sure you're tired of me repeating this by now, but would love to understand why some of these are commented out. If these aren't needed, perhaps we just delete the files completely.

File: server/src/patches/withFilter.ts [hunk 4]
```diff
@@ -1,27 +0,0 @@
...
-
-type ResolverFn<TResult, TParent, TContext, TArgs> = (
-  parent: TParent,
-  args: TArgs,
-  context: TContext,
-  info: GraphQLResolveInfo
-) => Promise<TResult> | TResult;
-
-// https://github.com/apollographql/graphql-subscriptions/issues/261#issuecomment-1246300357
-export const withFilter = withFilterOriginal as unknown as <
```
Comment (jerelmiller) [inline server/src/patches/withFilter.ts:19]: Were you able to get subscription types working without this type override?

File: server/src/resolvers/default.ts [hunk 5]
```diff
@@ -1,6 +0,0 @@
-import { defaultFieldResolver } from 'graphql';
-import { wrapWithSynthetics } from './helpers';
-
-const defaultResolver = wrapWithSynthetics(defaultFieldResolver);
```
Comment (jerelmiller) [inline server/src/resolvers/default.ts:4]: Would love to maintain this behavior as well if we can to ensure we can artificially slow down any field in the schema.

File: subgraphs/playback/src/utils/schema.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+export const shouldLoadFullArtist = (
+  pathToArtist: string[],
+  info: GraphQLResolveInfo
+) => {
+  return ['followers', 'genres', 'images', 'popularity'].some((fieldName) =>
+    selectsField([...pathToArtist, fieldName], info)
+  );
+};
+
+export const selectsField = (path: string[], info: GraphQLResolveInfo) => {
```
Comment (jerelmiller) [inline subgraphs/playback/src/utils/schema.ts:12]: Looks like these functions are duplicates of [`src/utils/graphql.ts`](https://github.com/apollographql/spotify-showcase/pull/66/files?diff=unified&w=1#diff-d90aba0bad98da6a35018e48b47bd5b1b3aa619bb9192e3263d116bb9d1e4ab6). Any chance we could just use those and remove from this file?

File: subgraphs/spotify/schema.graphql [hunk 7]
```diff
@@ -1,3 +1,22 @@
...
+  "URL where the subgraph's owner can be reached"
+  url: String
+  "Other relevant notes can be included here; supports markdown links"
+  description: String
+) on SCHEMA
+
+extend schema
+  @contact(
+    name: "Spotify Playback Subscription Server"
+    url: "https://myteam.slack.com/archives/teams-chat-room-url"
```
Comment (jerelmiller) [inline subgraphs/spotify/schema.graphql:15]: Should these be filled out in a meaningful way?

File: subgraphs/spotify/src/@types/graphql.d.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+declare module '*.graphql' {
```
Comment (jerelmiller) [inline subgraphs/spotify/src/@types/graphql.d.ts:1]: Since you switched to `readFileSync` to get the type defs, you might not need this anymore if you're not importing `.graphql` files.

File: subgraphs/spotify/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
+import { readFileSync } from "fs";
+import gql from "graphql-tag";
+import { buildSubgraphSchema } from "@apollo/subgraph";
+import { ApolloServer, GraphQLResponse } from "@apollo/server";
+import { expressMiddleware } from "@apollo/server/express4";
+import resolvers from "./resolvers";
+import { ContextValue } from "./types/ContextValue";
+const port = process.env.PORT ?? "4001";
+const subgraphName = require("../package.json").name;
```
Comment (jerelmiller) [inline subgraphs/spotify/src/index.ts:9]: Out of curiosity, any reason we can't use `import` here?

File: subgraphs/spotify/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                }),
+                document: parse(query),
+                contextValue,
+                variableValues: variables,
+                operationName,
+              });
+
+              return {
+                http: request.http,
+                body: { kind:"single", singleResult: response },
```
Comment (jerelmiller) [inline subgraphs/spotify/src/index.ts:64]: Looks like we might need to run prettier on this file :)

File: subgraphs/spotify/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+            },
+          };
+        },
+      },
+    ],
+  });
+
+  await server.start();
+
+  app.use(express.static("public"));
```
Comment (jerelmiller) [inline subgraphs/spotify/src/index.ts:75]: Same here, we might not need this anymore.

File: subgraphs/spotify/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+          },
+          mock: token ? false : true
+        };
+      },
+    })
+  );
+
+  await new Promise<void>((resolve) => httpServer.listen({ port }, resolve));
+
+  console.log(`🚀 Subscription endpoint ready at ws://localhost:${port}`);
```
Comment (jerelmiller) [inline subgraphs/spotify/src/index.ts:100]: This subgraph isn't subscription based correct? Perhaps we change this to:

```suggestion
  console.log(`🚀 GraphQL endpoint ready at http://localhost:${port}`);
```

File: subgraphs/spotify/src/resolvers/Developer.ts [hunk 13]
```diff
@@ -0,0 +1,6 @@
+import { getFieldConfigs } from '../fieldConfigs';
+import { DeveloperResolvers } from '../__generated__/resolvers-types';
+
+export const Developer: DeveloperResolvers = {
```
Comment (jerelmiller) [inline subgraphs/spotify/src/resolvers/Developer.ts:4]: Eventually this stuff might make sense in its own subgraph 🙂. This is specifically for our version of the app where we can add synthetic timeouts and errors. This stuff is not available in the Spotify API as its specific to this demo app.

No need to worry about it in this PR since its already massive, but something to consider for a future PR.

File: subgraphs/spotify/src/resolvers/PlaybackItem.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  __resolveType: (playbackItem) => {
+    switch (playbackItem.type) {
+      case "episode":
+        return "Episode";
+      case "track":
+        return "Track";
+    }
+  },
+  async __resolveReference(playbackItem, { dataSources }) {
+    //This is scritcly for a demo, not a good practice
```
Comment (jerelmiller) [inline subgraphs/spotify/src/resolvers/PlaybackItem.ts:13]:
```suggestion
    // This is strictly for a demo, not a good practice
```

Is this comment related to the `try/catch` or the `id`? I'm guessing for the try/catch. Any chance that `playbackItem` has a `type` property on it that would allow us to avoid the potential double fetch?

File: subgraphs/spotify/src/utils/constants.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+  'user-read-currently-playing',
+  'user-read-playback-position',
+  'user-read-playback-state',
+  'user-library-read',
+  'user-library-modify',
+  'user-top-read',
+  'user-read-recently-played',
+] as const;
+
+export const TOPICS = {
```
Comment (jerelmiller) [inline subgraphs/spotify/src/utils/constants.ts:20]: Are these needed in this subgraph since these are used for the subscription?

**Review by jerelmiller — CHANGES_REQUESTED** [review]
Hopefully this helps! The big reason I'm "blocking" is because I'd like to see the synthetic timeouts/errors added back to the resolvers which is super useful to understand how Suspense works.

File: client/codegen.ts [hunk 16]
```diff
@@ -0,0 +1,39 @@
+import type { CodegenConfig } from '@graphql-codegen/cli';
+
+const config: CodegenConfig = {
```
[the remark this answers is not in the document]
Comment (michael-watson) [reply client/codegen.ts:3]: Ya but that won't work for what is being auto-deployed in our infra. Railway and Netlify look at the specific folders for changes and deploy those folders, so if the codegen isn't split apart, the deployment won't have the codegen config or we have to add some copy step as part of the deployment build process.

File: client/schema.graphql
[the remark this answers is not in the document]
Comment (michael-watson) [reply client/schema.graphql:1]: Good catch, I forgot to add this to our codegen script to get the schema from GraphOS (no need to copy anything)

`"codegen": "rover graph fetch Spotify-tb7du2@prod-external > schema.graphql && graphql-codegen"`

File: subgraphs/playback/src/resolvers/Episode.ts [hunk 17]
```diff
@@ -0,0 +1,20 @@
+// import { EpisodeResolvers } from '../__generated__/resolvers-types';
```
[the remark this answers is not in the document]
Comment (michael-watson) [reply subgraphs/playback/src/resolvers/Episode.ts:1]: These don't exist in the `playback` subgraph anymore, I'll delete the files

File: subgraphs/playback/src/resolvers/PlaylistTrack.ts
[the remark this answers is not in the document]
Comment (michael-watson) [reply subgraphs/playback/src/resolvers/PlaylistTrack.ts:1]: These don't exist in the `playback` subgraph anymore, I'll delete the files

File: client/src/types/api.ts [hunk 18]
```diff
@@ -1085,6 +1085,7 @@ export type Query = {
   show: Maybe<Show>;
   /** Get Spotify catalog information for several shows based on their Spotify IDs. */
   shows: Maybe<Array<Show>>;
+  test: Maybe<Scalars['String']['output']>;
```
[the remark this answers is not in the document]
Comment (michael-watson) [reply client/src/types/api.ts:1088]: Good catch, fixed the schema

File: client/src/types/api.ts [hunk 19]
```diff
@@ -2355,9 +2356,9 @@ export type CurrentUserQueryVariables = Exact<{ [key: string]: never; }>;
 
 export type CurrentUserQuery = { me: { __typename: 'CurrentUser', user: { __typename: 'User', id: string, displayName: string | null, images: Array<{ __typename: 'Image', url: string }> | null } } | null };
 
-export type DevicePopover_playbackState = { __typename: 'PlaybackState', isPlaying: boolean, device: { __typename: 'Device', id: string } };
+export type DevicePopover_playbackState = { __typename: 'PlaybackState', isPlaying: boolean, device: { __typename: 'Device', id: string | null } };
 
-export type DevicePopover_devices = { __typename: 'Device', id: string, name: string, type: string };
+export type DevicePopover_devices = { __typename: 'Device', id: string | null, name: string, type: string };
```
[the remark this answers is not in the document]
Comment (michael-watson) [reply client/src/types/api.ts:2361]: I don't remember why I made this change or how it happened, but reverted and didn't have issues

File: subgraphs/playback/src/utils/schema.ts (hunk 6 above)
Comment (michael-watson) [reply subgraphs/playback/src/utils/schema.ts:12]: Yes...I don't even know why that file is there as it isn't referenced anywhere 🙃

File: subgraphs/spotify/src/index.ts (hunk 9 above)
Comment (michael-watson) [reply subgraphs/spotify/src/index.ts:9]: Copy/pasta accident. We're not using that anyways so I went ahead and removed it

File: subgraphs/spotify/src/resolvers/Developer.ts (hunk 13 above)
Comment (michael-watson) [reply subgraphs/spotify/src/resolvers/Developer.ts:4]: Good callout, I created #67 to track that and make it happen. I think it's an easy add on with another Netlify function.

File: subgraphs/spotify/src/utils/constants.ts (hunk 15 above)
Comment (michael-watson) [reply subgraphs/spotify/src/utils/constants.ts:20]: Nope, deleted

File: subgraphs/spotify/src/utils/schema.ts [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+export const shouldLoadFullArtist = (
+  pathToArtist: string[],
+  info: GraphQLResolveInfo
+) => {
+  return ['followers', 'genres', 'images', 'popularity'].some((fieldName) =>
+    selectsField([...pathToArtist, fieldName], info)
+  );
+};
+
+export const selectsField = (path: string[], info: GraphQLResolveInfo) => {
```
[the remark this answers is not in the document]
Comment (michael-watson) [reply subgraphs/spotify/src/utils/schema.ts:12]: yup!

File: subgraphs/playback/src/resolvers/helpers.ts (hunk 2 above)
Comment (michael-watson) [reply subgraphs/playback/src/resolvers/helpers.ts:23]: Added this back in, just had to make some small tweaks!

File: subgraphs/spotify/schema.graphql (hunk 7 above)
Comment (michael-watson) [reply subgraphs/spotify/schema.graphql:15]: Yup, added pointers to our Discord

File: server/src/patches/withFilter.ts (hunk 4 above)
Comment (michael-watson) [reply server/src/patches/withFilter.ts:19]: yup!

File: codegen.ts
[the remark this answers is not in the document]
Comment (michael-watson) [reply codegen.ts:1]: Done!

File: client/codegen.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (michael-watson) [reply client/codegen.ts:3]: We talked about having the codegen files be committed and not on every build. This is resolved

File: client/src/components/CurrentUserMenu.tsx [hunk 21]
```diff
@@ -41,7 +41,7 @@ const CurrentUserMenu = () => {
       </DropdownMenu.Trigger>
       <DropdownMenu.Menu align="end">
         <DropdownMenu.Item to="/settings">Settings</DropdownMenu.Item>
-        <DropdownMenu.Item to={`${import.meta.env.VITE_SERVER_HOST}/graphql`}>
+        <DropdownMenu.Item to={`https://studio.apollographql.com/public/spotify-ev3of9/variant/prod/home`}>
```
[the remark this answers is not in the document]
Comment (jerelmiller) [reply client/src/components/CurrentUserMenu.tsx]: In that case, wouldn't we want to just set the `VITE_SERVER_HOST` environment variable to `https://studio.apollographql.com/public/spotify-ev3of9/variant/prod/home` in production? We'd probably want to move the `/graphql` path into the enviroment variable itself in that case, but at least this way we should be able to control per-environment via the env variable what this value should be. Thoughts?

File: client/codegen.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (jerelmiller) [reply client/codegen.ts]: Chatted in Slack about this change.

File: subgraphs/playback/src/resolvers/Mutation.ts [hunk 22]
```diff
@@ -18,22 +18,47 @@ const refreshPlaybackState = async (
...
+          offset: maybeDeep(input?.offset),
+          position_ms: maybe(input?.positionMs),
+          uris: maybe(input?.uris),
+        },
+        params: { device_id: maybe(input?.deviceId) },
+      });
+      resumed = true;
+    } catch (err) {
+      console.log(err);
+      if (err?.extensions?.reason == "NO_ACTIVE_DEVICE") {
```
Comment (jerelmiller) [inline subgraphs/playback/src/resolvers/Mutation.ts:35]: 🙌

**Comment by jerelmiller** [comment]
Looks like you might need to run `codegen` one more time and check in the changes to get ESLint to pass.

**Review by jerelmiller — APPROVED** [review]
I'm so so so stoked about this! Thanks for putting in all this work to make this a part of GraphOS! I'm also looking forward to trying out `@defer` in this app somewhere since we're now using the router :)
